File: backend/src/services/tools/voxel_debug/voxel_debug.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <cstdint>
#include <cstring>
#include "../voxel_downsample/include/ankerl/unordered_dense.h"

// Fast hash function for 64-bit integers (matches Rust's FxHash exactly)
struct FastHash {
    size_t operator()(uint64_t x) const noexcept {
        constexpr uint64_t K = 0x517cc1b727220a95ULL;
        x = x * K;
        x = (x << 5) | (x >> 59);
        return static_cast<size_t>(x);
    }
};

// Binary protocol for fast I/O (replaces text I/O)
// Input format: [uint32_t pointCount][float voxelSize][float minX][float minY][float minZ][float maxX][float maxY][float maxZ][float* pointData]
// Output format: [uint32_t voxelCount][float* voxelGridPositions]

int main() {
    // OPTIMIZATION: Read binary input instead of text (much faster!)
    // Binary format: [uint32_t pointCount][float voxelSize][float minX][float minY][float minZ][float maxX][float maxY][float maxZ][float* pointData]
    
    // Read binary header (32 bytes: 4 for uint32 + 7*4 for floats)
    uint32_t pointCount;
    float voxelSize, minX, minY, minZ, maxX, maxY, maxZ;
    
    if (!std::cin.read(reinterpret_cast<char*>(&pointCount), sizeof(uint32_t))) {
        return 1; // Failed to read header
    }
    if (!std::cin.read(reinterpret_cast<char*>(&voxelSize), sizeof(float))) {
        return 1;
    }
    if (!std::cin.read(reinterpret_cast<char*>(&minX), sizeof(float))) {
        return 1;
    }
    if (!std::cin.read(reinterpret_cast<char*>(&minY), sizeof(float))) {
        return 1;
    }
    if (!std::cin.read(reinterpret_cast<char*>(&minZ), sizeof(float))) {
        return 1;
    }
    if (!std::cin.read(reinterpret_cast<char*>(&maxX), sizeof(float))) {
        return 1;
    }
    if (!std::cin.read(reinterpret_cast<char*>(&maxY), sizeof(float))) {
        return 1;
    }
    if (!std::cin.read(reinterpret_cast<char*>(&maxZ), sizeof(float))) {
        return 1;
    }
    
    // Validate input
    if (pointCount == 0 || voxelSize <= 0) {
        // Write empty result (4 bytes: voxelCount = 0)
        uint32_t voxelCount = 0;
        std::cout.write(reinterpret_cast<const char*>(&voxelCount), sizeof(uint32_t));
        std::cout.flush();
        return 0;
    }
    
    // Read point data directly into vector (zero-copy where possible)
    const size_t floatCount = pointCount * 3;
    std::vector<float> pointCloudData(floatCount);
    
    if (!std::cin.read(reinterpret_cast<char*>(pointCloudData.data()), floatCount * sizeof(float))) {
        return 1; // Failed to read point data
    }
    
    float* inputData = pointCloudData.data();
    
    // ULTRA OPTIMIZED C++ voxel debug algorithm - matches WASM version
    // OPTIMIZATION 1: Pre-calculate ALL constants at the start (like Rust/WASM does)
    float invVoxelSize = 1.0f / voxelSize;
    float halfVoxelSize = voxelSize * 0.5f;
    float offsetX = minX + halfVoxelSize;
    float offsetY = minY + halfVoxelSize;
    float offsetZ = minZ + halfVoxelSize;
    
    // OPTIMIZATION 2: Use ankerl::unordered_dense::set with FastHash (fast hash set, matches Rust FxHash performance)
    // Don't reserve - let it grow naturally like Rust/WASM (reserve can cause overhead if estimate is wrong)
    ankerl::unordered_dense::set<uint64_t, FastHash> voxelKeys;
    
    // OPTIMIZATION 3: Process points in chunks for better cache locality
    const int CHUNK_SIZE = 1024;
    for (uint32_t chunkStart = 0; chunkStart < pointCount; chunkStart += CHUNK_SIZE) {
        uint32_t chunkEnd = std::min(chunkStart + CHUNK_SIZE, pointCount);
        
        for (uint32_t i = chunkStart; i < chunkEnd; i++) {
            int i3 = i * 3;
            float x = inputData[i3];
            float y = inputData[i3 + 1];
            float z = inputData[i3 + 2];
            
            // OPTIMIZATION 4: Use floor and multiplication instead of division
            int voxelX = static_cast<int>(std::floor((x - minX) * invVoxelSize));
            int voxelY = static_cast<int>(std::floor((y - minY) * invVoxelSize));
            int voxelZ = static_cast<int>(std::floor((z - minZ) * invVoxelSize));
            
            // OPTIMIZATION 5: Create integer hash key (same as WASM)
            uint64_t voxelKey = (static_cast<uint64_t>(voxelX) << 32) |
                               (static_cast<uint64_t>(voxelY) << 16) |
                               static_cast<uint64_t>(voxelZ);
            
            // OPTIMIZATION 6: Store unique voxel keys only (same as WASM)
            // Use emplace to avoid copy (construct in-place)
            voxelKeys.emplace(voxelKey);
        }
    }
    
    // OPTIMIZATION 7: Pre-allocate result vector
    std::vector<float> voxelGridPositions;
    voxelGridPositions.resize(voxelKeys.size() * 3); // Pre-allocate exact size
    
    // OPTIMIZATION 8: Single pass conversion with pointer arithmetic (like WASM)
    // Use pointer arithmetic for maximum performance (like optimized voxel downsampling)
    float* outputPtr = voxelGridPositions.data();
    
    for (const uint64_t voxelKey : voxelKeys) {
        // Extract voxel coordinates from integer key (optimized bit operations, same as WASM)
        // voxelX: bits 32-63 (32 bits), voxelY: bits 16-31 (16 bits), voxelZ: bits 0-15 (16 bits)
        int voxelX = static_cast<int32_t>(voxelKey >> 32);
        int voxelY = static_cast<int16_t>((voxelKey >> 16) & 0xFFFF); // Sign-extend 16-bit to int
        int voxelZ = static_cast<int16_t>(voxelKey & 0xFFFF); // Sign-extend 16-bit to int
        
        // Calculate voxel grid position (center of voxel grid cell) - direct calculation and write
        *outputPtr++ = offsetX + static_cast<float>(voxelX) * voxelSize;
        *outputPtr++ = offsetY + static_cast<float>(voxelY) * voxelSize;
        *outputPtr++ = offsetZ + static_cast<float>(voxelZ) * voxelSize;
    }
    
    // OPTIMIZATION 10: Write binary output instead of text (much faster!)
    // Binary format: [uint32_t voxelCount][float* voxelGridPositions]
    uint32_t voxelCount = static_cast<uint32_t>(voxelKeys.size());
    std::cout.write(reinterpret_cast<const char*>(&voxelCount), sizeof(uint32_t));
    std::cout.write(reinterpret_cast<const char*>(voxelGridPositions.data()), voxelGridPositions.size() * sizeof(float));
    std::cout.flush();
    
    return 0;
}

File: frontend/src/wasm/rust/src/voxel_debug.rs
use rustc_hash::FxHashSet;

pub fn generate_voxel_centers_internal(
    points: &[f32],
    voxel_size: f32,
    min_x: f32,
    min_y: f32,
    min_z: f32,
) -> Vec<f32> {
    // Pre-calculate constants to avoid repeated calculations
    let inv_voxel_size = 1.0 / voxel_size;
    let half_voxel_size = voxel_size * 0.5;
    let offset_x = min_x + half_voxel_size;
    let offset_y = min_y + half_voxel_size;
    let offset_z = min_z + half_voxel_size;
    
    // Use fast hash set with integer keys to track unique voxels
    let mut voxel_keys: FxHashSet<u64> = FxHashSet::default();
    
    // Process points in chunks for better CPU cache performance
    const CHUNK_SIZE: usize = 1024;
    let point_count = points.len() / 3;
    
    for chunk_start in (0..point_count).step_by(CHUNK_SIZE) {
        let chunk_end = (chunk_start + CHUNK_SIZE).min(point_count);
        
        for i in chunk_start..chunk_end {
            let i3 = i * 3;
            let x = points[i3];
            let y = points[i3 + 1];
            let z = points[i3 + 2];
            
            // Calculate voxel grid coordinates using multiplication (faster than division)
            let voxel_x = ((x - min_x) * inv_voxel_size).floor() as i32;
            let voxel_y = ((y - min_y) * inv_voxel_size).floor() as i32;
            let voxel_z = ((z - min_z) * inv_voxel_size).floor() as i32;
            
            // Combine coordinates into single integer hash key
            let voxel_key = ((voxel_x as u64) << 32) | ((voxel_y as u64) << 16) | (voxel_z as u64);
            
            voxel_keys.insert(voxel_key);
        }
    }
    
    // Pre-allocate result vector with exact capacity
    let voxel_count = voxel_keys.len();
    let mut centers = Vec::with_capacity(voxel_count * 3);
    
    // Convert unique voxel keys to center positions
    for voxel_key in voxel_keys {
        // Extract voxel coordinates from integer key
        let voxel_x = (voxel_key >> 32) as i32;
        let voxel_y = ((voxel_key >> 16) & 0xFFFF) as i16 as i32;
        let voxel_z = (voxel_key & 0xFFFF) as i16 as i32;
        
        // Calculate voxel center position
        let center_x = offset_x + voxel_x as f32 * voxel_size;
        let center_y = offset_y + voxel_y as f32 * voxel_size;
        let center_z = offset_z + voxel_z as f32 * voxel_size;
        
        centers.push(center_x);
        centers.push(center_y);
        centers.push(center_z);
    }
    
    centers
}

